### PR TITLE
[cloudbuild] Enable parallel builds

### PR DESCRIPTION
**Pull request by jprzychodzen**
* Add two flags, --enable_parallel and --force_parallel
* Change is backward compatibile

### CODE REVIEW

File: versioning/scripts/cloudbuild/main.go [hunk 1]
```diff
@@ -251,6 +296,8 @@ func main() {
 	newTagsPtr := flag.Bool("new_tags", false, "Require that image tags do not already exist.")
 	firstTagOnly := flag.Bool("first_tag", false, "Build only the first per version.")
 	timeoutPtr := flag.Int("timeout", 0, "Timeout in seconds. If not set, the default Cloud Build timeout is used.")
+	enableParallel := flag.Bool("enable_parallel", false, "Enable parallel build and bigger VM")
```
Comment (huyhg) [inline versioning/scripts/cloudbuild/main.go]: shouldn't this be `true` by default? we still respect what's in the config file right, which by default is off

File: versioning/scripts/cloudbuild/main.go [hunk 2]
```diff
@@ -129,11 +159,25 @@ type imageBuildTemplateData struct {
 
 type cloudBuildTemplateData struct {
 	RequireNewTags bool
+	Parallel       bool
 	ImageBuilds    []imageBuildTemplateData
 	AllImages      []string
 	TimeoutSeconds int
 }
 
+func parallelize(
```
Comment (huyhg) [inline versioning/scripts/cloudbuild/main.go]: rename to `shouldParallelize`

File: versioning/scripts/cloudbuild/main.go [hunk 3]
```diff
@@ -129,11 +159,25 @@ type imageBuildTemplateData struct {
...
+	options cloudBuildOptions,
```
Comment (huyhg) [inline versioning/scripts/cloudbuild/main.go]: all args should just be on one line

File: versioning/scripts/cloudbuild/main.go [hunk 4]
```diff
@@ -54,6 +61,10 @@ const cloudBuildTemplateString = `steps:
       - 'build'
       - '--tag={{ .Tag }}'
       - '{{ .Directory }}'
+{{- if (eq $parallel true) }}
```
Comment (huyhg) [inline versioning/scripts/cloudbuild/main.go]: should these just be `if $parallel`?

Comment (jprzychodzen) [reply versioning/scripts/cloudbuild/main.go]: Done.

I was trying to do something like that at beginning, but it failed. Now I cannot check difference.

File: versioning/scripts/cloudbuild/main.go (hunk 2 above)
Comment (jprzychodzen) [reply versioning/scripts/cloudbuild/main.go]: Done.

File: versioning/scripts/cloudbuild/main.go (hunk 3 above)
Comment (jprzychodzen) [reply versioning/scripts/cloudbuild/main.go]: Done.

File: versioning/scripts/cloudbuild/main.go (hunk 1 above)
Comment (jprzychodzen) [reply versioning/scripts/cloudbuild/main.go]: I do want to keep this change as optional, with backward compatibility. It requires tests to parallelable, so this change can possibly break builds that depend on sequential execution.

Comment (huyhg) [reply versioning/scripts/cloudbuild/main.go]: I understand this as: even when this flag is `true`, unless we change the config files, no parallelization will automatically happen. `forceParallel` on the other hand will break non-ready tests.

File: versioning/scripts/cloudbuild/main.go [hunk 5]
```diff
@@ -165,10 +165,8 @@ type cloudBuildTemplateData struct {
 	TimeoutSeconds int
 }
 
-func parallelize(
-	options cloudBuildOptions,
-	numberOfVersions int,
-	numberOfTests int) bool {
+func shouldParallelize(
+	options cloudBuildOptions, numberOfVersions int, numberOfTests int) bool {
```
Comment (huyhg) [inline versioning/scripts/cloudbuild/main.go]: to previous line as well

File: versioning/scripts/cloudbuild/main.go (hunk 1 above)
Comment (jprzychodzen) [reply versioning/scripts/cloudbuild/main.go]: What kind of config file you have in mind? I do not use any. Please tell me more about this feature.

In this patch logic is as follows:
- no flag - as is, serial build,
- enable flag - if 'big repo', then parallel build, else serial build,
- force flag or both flags - parallel build.

Changing enable_parallel to default true:
- no flag - if 'big repo', then parallel build, else serial build,
- negate enable flag - as is, serial build,
- force flag - parallel build.

File: versioning/scripts/cloudbuild/main.go (hunk 5 above)
Comment (jprzychodzen) [reply versioning/scripts/cloudbuild/main.go]: Done.

File: versioning/scripts/cloudbuild/main.go (hunk 1 above)
Comment (huyhg) [reply versioning/scripts/cloudbuild/main.go]: My bad--for whatever reason I thought our `versions.yaml` has a bit indicating whether it desires to be parallelized. I see that 'big repo' is defined in `shouldParallelize`.

**Pull request merged**